doc/crdt-details.md

```diff
@@ -0,0 +1,1086 @@
+
+```rust
+/// Find an index before which everything is the same
+fn find_base_index(a: &[Revision], b: &[Revision]) -> usize {
+    assert!(a.len() > 0 && b.len() > 0);
+    assert!(a[0].rev_id == b[0].rev_id);
+    // TODO find the maximum base revision.
+    // returning 1 should have the same behavior, but worse performance
+    return 1;
+}
+
+/// Find a set of revisions common to both lists
+fn find_common(a: &[Revision], b: &[Revision]) -> BTreeSet<RevId> {
+    // TODO make this faster somehow?
+    let a_ids: BTreeSet<RevId> = a.iter().map(|r| r.rev_id).collect();
+    let b_ids: BTreeSet<RevId> = b.iter().map(|r| r.rev_id).collect();
+    a_ids.intersection(&b_ids).cloned().collect()
+}
+```
+
+#### Rearranging
+
+To make the rest of the algorithm easier, it would be nice if we didn't have to worry about any of the common revisions and could just work with the new revisions on each side. We can realize this by transforming and reordering all the new revisions on each side after all of the common revisions.
+
+We can safely use transforms to reorder two revisions if neither of the revisions depends on the other. A revision depends on all the revisions that were in the `Engine` at the time `Engine::edit_rev` was called. We know that a new revision can't depend on common revision after it, because the append-only nature means that ordering couldn't occur. Furthermore, the common revision can't depend on the new revision, because by definitions the other peer has the common revision but not the new revision, and merges never merge a revision but not its dependencies. Thus, even without explicit dependency information, we know we can reorder all the new revisions on each side after all the common revisions.
+
+Note that though it's nice to think about it as reordering the histories, really we just need to compute the transformed new revisions, transforming each new revision to be based on all common revisions after it. See the "computed" and "taken to next step" annotations of the diagram below, note that transformed operations are marked with a prime tick but they maintain their ID.
+
+![CRDT merge rearrange](img/merge-rearrange.png)
+
+We do the rearranging by working from the end of the list of revisions to the beginning. We build up a `Subset` representing all the characters that were added to the union string by the common revisions after a point. Starting with the identity transformation (an entirely 0-count `Subset`), whenever we encounter a common revision we add the characters it inserts to the transform. When we encounter a new revision we `transform_expand` its inserts and deletes by the transform `Subset` and add it to the output `Vec`. Then we `transform_shrink` the transform `Subset` by the characters inserted in the new revision, because we only want to fast-forward new revisions over common revisions after them, not new revisions after them.
```

Comment:
"after a point.": Which point? an arbitrary point? This is unclear.

Also I might rephrase the first sentence, and just say "we do the rearranging by working backwards from the end of the list of revisions."